### DISCUSSION TITLE
orml_unknown_tokens can be target for spam?

### DISCUSSION THREAD

**Discussion by dzmitry-lahoda** [open-web3-stack/open-runtime-module-library#853] [Q&A] [answered]
Using `MultiCurrencyAdapter` to send assets to `orml_unknown_tokens::concrete_fungible_balances` leads to creating accounts per token per recipient.

So one may send a minimal fee as payment for the first payment asset (as I see there must be paid or unknown tokens lost?). 

Can send second assets with random id and random recipient and use big storage because ED is not taken for this. 

What can be done?

For example, store assets on the sender chain account and restore assets using historical data from somewhere (not automatical).

Of may take ED from asset fee ?

What happens if the sender sends a token and no fee is provided at all? Do unknown tokens fail fully?  Could at least store assets on account if chain sender?

**Comment by xlc** [comment, marked as answer]
It is a known issue https://github.com/paritytech/polkadot/issues/2702

**Comment by shortcut119** [comment]
pub struct SafeMultiCurrencyAdapter<T>(PhantomData<T>);

impl<T: Config> MultiCurrency for SafeMultiCurrencyAdapter<T> {
    fn deposit(currency_id, who, amount) -> DispatchResult {
        if is_unknown(currency_id) {
            ensure!(fee_paid(who)?, Error::<T>::MissingFee);
        }
        // then forward to normal deposit
    }
}
